packages/evm/contracts/payments/PaymentsBook.sol

```diff
@@ -0,0 +1,535 @@
+        if (randomness == bytes32(0)) {
+            randomness = blockhash(getBlockNumber() - 1);
+        }
+
+        // Use randomness to choose a Court term of the requested period and query the total amount of juror tokens active at that term
+        IJurorsRegistry jurorsRegistry = _jurorsRegistry();
+        uint64 periodBalanceCheckpoint = periodStartTermId.add(uint64(uint256(randomness) % periodDuration));
+        totalActiveBalance = jurorsRegistry.totalActiveBalanceAt(periodBalanceCheckpoint);
+
+        _period.balanceCheckpoint = periodBalanceCheckpoint;
+        _period.totalActiveBalance = totalActiveBalance;
+        return (periodBalanceCheckpoint, totalActiveBalance);
+    }
+
+    /**
+    * @dev Internal function to set a new governor share value
+    * @param _governorSharePct New permyriad of collected fees that will be allocated to the governor of the Court (‱ - 1/10,000)
+    */
+    function _setGovernorSharePct(uint16 _governorSharePct) internal {
+        // Check governor share is not greater than 10,000‱
+        require(PctHelpers.isValid(_governorSharePct), ERROR_OVERRATED_GOVERNOR_SHARE_PCT);
+
+        emit GovernorSharePctChanged(governorSharePct, _governorSharePct);
+        governorSharePct = _governorSharePct;
+    }
+
+    /**
+    * @dev Internal function to tell the identification number of the current period
+    * @return Identification number of the current period
+    */
+    function _getCurrentPeriodId() internal view returns (uint256) {
+        // Since the Court starts at term #1, and the first payment period is #0, then subtract one unit to the current term of the Court
+        uint64 termId = _getCurrentTermId();
+        require(termId > 0, ERROR_COURT_HAS_NOT_STARTED);
+
+        // No need for SafeMath: we already checked that the term ID is at least 1
+        uint64 periodId = (termId - START_TERM_ID) / periodDuration;
+        return uint256(periodId);
+    }
+
+    /**
+    * @dev Internal function to get the current period
+    * @return periodId Identification number of the current period
+    * @return period Current period instance
+    */
+    function _getCurrentPeriod() internal view returns (uint256 periodId, Period storage period) {
+        periodId = _getCurrentPeriodId();
+        period = periods[periodId];
+    }
+
+    /**
+    * @dev Internal function to get the Court term in which a certain period starts
+    * @param _periodId Identification number of the period querying the start term of
+    * @return Court term where the given period starts
+    */
+    function _getPeriodStartTermId(uint256 _periodId) internal view returns (uint64) {
+        // Periods are measured in Court terms. Since Court terms are represented in uint64, we are safe to use uint64 for period ids too.
+        // We are using SafeMath here because if any user calls `getPeriodBalanceDetails` for a huge period ID,
```

Comment:
I don't think this comment is still valid; `getPeriodBalanceDetails()` doesn't call `_getPeriodStartTermId()`.

However, while it's not that big of a problem, you can technically cause this to always fail if you set a large enough `periodDuration` on initialization (e.g. UINT64_MAX). One way around this is to double check a multiplication against the value, e.g. 1000k, works when setting the period duration.

Reply:
I'll update the comment but I don't it's extremely necessary to have a check for the `periodDuration`, whatever number you use as a benchmark for the multiplication will sound like you cannot have more than that number of periods, which at the end will always depend on the term duration and period duration you're using

Reply:
Right, that makes sense. You'll never be able to claim shares, but that's because the first period hasn't ended (and won't ever end).